### PR TITLE
Shutdown hooks

### PR DESCRIPTION
**Pull request by jshook**
This is to unblock an important resource leak fix for driver adapters. Specifically it uses the internal Shutdownable decorator interface to enable auto-shutdown of resources. This will unblock a change needed for the Pulsar adapter.

### CODE REVIEW

File: adapter-cqld4/src/main/java/io/nosqlbench/adapter/cqld4/Cqld4Space.java [hunk 1]
```diff
@@ -299,4 +300,8 @@ public static NBConfigModel getConfigModel() {
 
     }
 
+    @Override
+    public void shutdown() {
+        this.getSession().close();
```
Comment (jeffbanks) [inline adapter-cqld4/src/main/java/io/nosqlbench/adapter/cqld4/Cqld4Space.java:305]: Should we consider logging an error or warn level message here?  I believe this implements auto-closeable but throws an exception.

File: nbr-examples/src/test/resources/scripts/examples/extension_histostatslogger.js [hunk 2]
```diff
@@ -18,16 +18,14 @@ activitydef = {
     "alias" : "testhistostatslogger",
     "driver" : "diag",
     "cycles" : "50000",
-    "threads" : "20",
-    "interval" : "2000",
-    "targetrate" : "10000.0",
+    "threads" : "5",
```
Comment (jeffbanks) [inline nbr-examples/src/test/resources/scripts/examples/extension_histostatslogger.js:21]: Looks like this one snuck in this PR, but makes sense as we wanted this anyway.

File: adapters-api/src/main/java/io/nosqlbench/engine/api/activityimpl/uniform/DriverSpaceCache.java [hunk 3]
```diff
@@ -64,4 +66,8 @@ public S get(String name) {
         return cache.computeIfAbsent(name, newSpaceFunction);
     }
 
+    public Map<String,S> getElements() {
+        return Collections.unmodifiableMap(cache);
```
Comment (jeffbanks) [inline adapters-api/src/main/java/io/nosqlbench/engine/api/activityimpl/uniform/DriverSpaceCache.java:70]: I like the unmodifiable usage.

**Review by jeffbanks — APPROVED** [review]
LGTM overall, just one comment for consideration.

File: adapter-cqld4/src/main/java/io/nosqlbench/adapter/cqld4/Cqld4Space.java (hunk 1 above)
Comment (jshook) [reply adapter-cqld4/src/main/java/io/nosqlbench/adapter/cqld4/Cqld4Space.java:305]: As in, wrap any exceptions from `.close()` and throw an error there? We're not using try-with-resources here, so I don't know if auto-closeable can help much.

Comment (jshook) [reply adapter-cqld4/src/main/java/io/nosqlbench/adapter/cqld4/Cqld4Space.java:305]: But I do agree that the code is not defensive enough without an exception handler.

Comment (jeffbanks) [reply adapter-cqld4/src/main/java/io/nosqlbench/adapter/cqld4/Cqld4Space.java:305]: Right, I was referring to the usage of the session instance itself and if `null` don't attempt to close, while wrapping the close() w/ try/catch to ensure we log out that it couldn't shutdown fully for some reason.

The session implements `AsyncAutoCloseable` so I was thinking about how the lifecyle may have been somewhat controlled prior to the `shutdown` call on this Space instance, but we never really know how that is used.  😄

Comment (jshook) [reply adapter-cqld4/src/main/java/io/nosqlbench/adapter/cqld4/Cqld4Space.java:305]: I'm reworking this code to re-use auto-closeable at the space level. Also adding integrated tests to ensure that errors are propagated correctly. After some reflection, I think we should not do anything special for errors here as we already have a configurable error handler for such.

Comment (jshook) [reply adapter-cqld4/src/main/java/io/nosqlbench/adapter/cqld4/Cqld4Space.java:305]: A nasty side-effect of using auto-closable is that developers are warned about not using try-with-resources. Switching this back to Shutdownable.

Comment (jshook) [reply adapter-cqld4/src/main/java/io/nosqlbench/adapter/cqld4/Cqld4Space.java]: ... except that this ^ only affects the core code, and so not a problem for devs. The latest changes use AutoCloseable and also keep the adapter generally stateless except for the spaces, so close() is not called through the adapter, but directly on the spaces visible to the activity. Added tests to provide coverage as well.

**Review by phact — APPROVED** [review]

**Pull request merged**